nexus/db-queries/src/db/datastore/resource_lock.rs

```diff
@@ -0,0 +1,184 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! [`DataStore`] methods on [`ResourceLock`]s.
+
+use super::DataStore;
+use crate::db;
+use crate::db::error::public_error_from_diesel_pool;
+use crate::db::error::ErrorHandler;
+use crate::db::error::TransactionError;
+use crate::db::model::ResourceLock;
+use async_bb8_diesel::AsyncConnection;
+use diesel::prelude::*;
+use omicron_common::api::external::CreateResult;
+use omicron_common::api::external::DeleteResult;
+use omicron_common::api::external::Error;
+use uuid::Uuid;
+
+impl DataStore {
+    /// Attempt locking a resource with a certain lock ID. If this function
+    /// returns Ok, then that lock was grabbed successfully.
+    pub async fn attempt_resource_lock(
+        &self,
+        resource_id: Uuid,
+        lock_id: Uuid,
+    ) -> CreateResult<()> {
+        self.pool()
+            .transaction(move |conn| {
+                use db::schema::resource_lock::dsl;
+
+                let existing_lock: Option<ResourceLock> = dsl::resource_lock
+                    .filter(dsl::resource_id.eq(resource_id))
+                    .first(conn)
+                    .optional()
+                    .map_err(|e| {
+                        public_error_from_diesel_pool(
+                            e.into(),
+                            ErrorHandler::Server,
+                        )
+                    })?;
+
+                if let Some(existing_lock) = existing_lock {
+                    if existing_lock.lock_id != lock_id {
+                        return Err(TransactionError::CustomError(
+                            Error::invalid_request(&format!(
+                            "resource {} already locked by another lock ID {}",
+                            resource_id, lock_id,
+                        )),
```

Comment:
With this current scheme, we have no idea "who" took the lock, correct? So in a situation where we see the resource is locked, but no progress is being made, we really have no choice but to wait forever, correct?

For example, suppose we do the following:

- Start saga node, in which we intend to grab the lock
- Try to grab the lock, succeed
- Before we wrote that we finished the saga node, we crash
- Reboot
- Re-read saga log, try to replay node, see that someone else grabbed the lock
- Fail, never succeed again, because the node that "Grabbed" the lock has crashed and no longer exists

Reply:
Oh, nevermind, I see that this is what the `lock_id` accomplishes - so if we see the same lock ID, we know it was our operation which triggered the action?